homeassistant/components/onvif/camera.py

```diff
@@ -374,155 +322,141 @@ async def async_obtain_input_uri(self):
-                        "Zoom": {"x": speed_val},
-                    }
-                    await self._ptz_service.GotoPreset(req)
-            except exceptions.ONVIFError as err:
-                if "Bad Request" in err.reason:
-                    self._ptz_service = None
-                    _LOGGER.debug("Camera '%s' doesn't support PTZ.", self._name)
-        else:
-            _LOGGER.debug("Camera '%s' doesn't support PTZ.", self._name)
+        pan_val = distance * PAN_FACTOR.get(pan, 0)
+        tilt_val = distance * TILT_FACTOR.get(tilt, 0)
+        zoom_val = distance * ZOOM_FACTOR.get(zoom, 0)
+        speed_val = speed
+        preset_val = preset
+        LOGGER.debug(
+            "Calling %s PTZ | Pan = %4.2f | Tilt = %4.2f | Zoom = %4.2f | Speed = %4.2f | Preset = %s",
+            move_mode,
+            pan_val,
+            tilt_val,
+            zoom_val,
+            speed_val,
+            preset_val,
+        )
+        try:
+            req = self._ptz_service.create_type(move_mode)
+            req.ProfileToken = self._profile_token
+            if move_mode == CONTINUOUS_MOVE:
+                req.Velocity = {
+                    "PanTilt": {"x": pan_val, "y": tilt_val},
+                    "Zoom": {"x": zoom_val},
+                }
+
+                await self._ptz_service.ContinuousMove(req)
+                await asyncio.sleep(continuous_duration)
+                req = self._ptz_service.create_type("Stop")
+                req.ProfileToken = self._profile_token
+                await self._ptz_service.Stop({"ProfileToken": req.ProfileToken})
+            elif move_mode == RELATIVE_MOVE:
+                req.Translation = {
+                    "PanTilt": {"x": pan_val, "y": tilt_val},
+                    "Zoom": {"x": zoom_val},
+                }
+                req.Speed = {
+                    "PanTilt": {"x": speed_val, "y": speed_val},
+                    "Zoom": {"x": speed_val},
+                }
+                await self._ptz_service.RelativeMove(req)
+            elif move_mode == ABSOLUTE_MOVE:
+                req.Position = {
+                    "PanTilt": {"x": pan_val, "y": tilt_val},
+                    "Zoom": {"x": zoom_val},
+                }
+                req.Speed = {
+                    "PanTilt": {"x": speed_val, "y": speed_val},
+                    "Zoom": {"x": speed_val},
+                }
+                await self._ptz_service.AbsoluteMove(req)
+            elif move_mode == GOTOPRESET_MOVE:
+                req.PresetToken = preset_val
+                req.Speed = {
+                    "PanTilt": {"x": speed_val, "y": speed_val},
+                    "Zoom": {"x": speed_val},
+                }
+                await self._ptz_service.GotoPreset(req)
+        except exceptions.ONVIFError as err:
+            if "Bad Request" in err.reason:
+                self._ptz_service = None
+                LOGGER.debug("Camera '%s' doesn't support PTZ.", self._name)
```

Comment:
Should we add an `else` here and log the error that happened to give feedback to the user?